File: recognize.py
from rich.progress import track
import os
import argparse
from typing import List, Callable, Any


def frame2ts(n, d):
    ms = d / 2 + (n - 1) * d
    h = int(ms / 1000 / 3600)
    ms -= h * 3600 * 1000
    m = int(ms / 1000 / 60)
    ms -= m * 60 * 1000
    s = int(ms / 1000)
    ms -= s * 1000
    ms = int(ms)
    return (
        str(h).zfill(2)
        + ":"
        + str(m).zfill(2)
        + ":"
        + str(s).zfill(2)
        + ","
        + str(ms).zfill(3)
    )


def recognize(
    add_args: Callable[[argparse.ArgumentParser], None],
    get_args: Callable[[dict[str, Any]], None],
    rec: Callable[[str], str],
):
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--img_dir", "-d", type=str, help="Sceenshot directory", required=True
    )
    parser.add_argument("--fps", type=str, help="Screenshot FPS", required=True)

    add_args(parser)

    args = vars(parser.parse_args())

    get_args(args)

    img_dir: str = args.get("img_dir")  # type: ignore
    fps: str = args.get("fps")  # type: ignore

    frame_duration = 1000 / float(fps)

    imgs: List[str] = []
    for item in filter(
        lambda x: x[0] != "." and str(x).endswith(".png"), os.listdir(img_dir)
    ):
        imgs.append(item)
    imgs.sort()

    subs: List[tuple[List[int], str]] = []

    for i in track(range(len(imgs))):
        file = os.path.join(img_dir, imgs[i])
        sentence = rec(file)
        print(imgs[i], sentence)
        if len(sentence) == 0:
            continue
        id = int(os.path.splitext(imgs[i])[0])
        if len(subs) > 0 and subs[-1][1] == sentence:
            subs[-1][0].append(id)
        else:
            subs.append(([id], sentence))

    print(subs)

    srt_items: List[str] = []
    for i, [ids, sentence] in enumerate(subs):
        st = frame2ts(ids[0], frame_duration)
        et = frame2ts(ids[-1], frame_duration)
        srt_items.append("\n".join([str(i + 1), st + " --> " + et, sentence]))

    srt_dest = img_dir + ".srt"

    with open(srt_dest, "w") as fp:
        for item in srt_items:
            fp.write("%s\n\n" % item)

    print("Saved to {}".format(srt_dest))

File: recognize.mjs
import { program } from "commander";
import fs from "fs/promises";
import { exec } from "child_process";
import path from "path";
import cliProgress from "cli-progress";

function frame2ts(n, d) {
  let ms = d / 2 + (n - 1) * d;
  const h = Math.floor(ms / 1000 / 3600);
  ms -= h * 3600 * 1000;
  const m = Math.floor(ms / 1000 / 60);
  ms -= m * 60 * 1000;
  const s = Math.floor(ms / 1000);
  ms -= s * 1000;
  return (
    `${h}`.padStart(2, "0") +
    ":" +
    `${m}`.padStart(2, "0") +
    ":" +
    `${s}`.padStart(2, "0") +
    "," +
    `${ms}`.padStart(3, "0")
  );
}

function execCommand(cmd) {
  return new Promise((resolve, reject) => {
    exec(cmd, (err, stdout, stderr) => {
      if (err) {
        reject(err);
      } else {
        resolve({
          stdout,
          stderr,
        });
      }
    });
  });
}

program
  .requiredOption("-d, --img_dir <string>", "Screenshot directory")
  .requiredOption("--fps <string>", "Screenshot FPS");

export default async function recognize(rec) {
  program.parse();
  const options = program.opts();
  const fps = +options.fps;
  if (isNaN(fps) || fps <= 0) {
    throw new Error("Invalid FPS");
  }
  const imgDir = options.img_dir;
  if (!(await fs.stat(imgDir)).isDirectory()) {
    throw new Error("Invalid image directory");
  }

  const imgs = (await fs.readdir(imgDir))
    .filter((item) => item.endsWith(".png"))
    .sort((a, b) => a - b);

  const subs = [];
  for (let i = 0; i < imgs.length; i++) {
    const img = imgs[i];
    const file = path.resolve(imgDir, imgs[i]);
    const sentence = await rec(file);
    console.log(img + "/" + imgs.length, sentence);
    if (sentence === "") {
      continue;
    }
    const id = +img.substring(0, img.length - path.extname(img).length);
    if (subs.length > 0 && subs[subs.length - 1].sentence === sentence) {
      subs[subs.length - 1].ids.push(id);
    } else {
      subs.push({
        ids: [id],
        sentence,
      });
    }
  }

  const frame_duration = 1000 / fps;

  const sub_items = [];
  subs.forEach(({ ids, sentence }, idx) => {
    const st = frame2ts(ids[0], frame_duration);
    const et = frame2ts(ids[ids.length - 1], frame_duration);
    sub_items.push(["" + idx, st + " --> " + et, sentence]);
  });

  const dest = imgDir + ".srt";

  await fs.writeFile(dest, sub_items.join("\n\n"));
  console.log(`Save to ${dest}`);
}

